_posts/2019-09-26-2019-building-admissions.md

```diff
@@ -0,0 +1,284 @@
+
+#### Sign-in button
+
+To kick off the auth flow for GitHub logins we need the user to click a link for the earlier request route we defined. To do that we added the following HTML to our `index.html.eex` file:
+
+```html
+<a class="button is-info is-medium" href="/auth/github">
+  <span class="icon">
+    <i class="fab fa-github"></i>
+  </span>
+  <span>Sign-in with GitHub</span>
+</a>
+```
+
+Now that our UI is updated we can call our Ueberauth integration code complete! The last step for us was setting up the application on GitHub. Once complete we pulled the `CLIENT_ID` and `CLIENT_SECRET` from the application settings and added them to our ENV.
+
+A user can now sign-in with a valid GitHub account. We need to handle the next step in the process: eligibility.
+
+### Verifying contributor status
+
+At this stage in the request our user has successfully authenticated with GitHub and now we need to determine if they've contributed to any of our repositories. To achieve this we need the GitHub API and the `token` that we in our OAuth response from GitHub. For this portion of the application the high level of what we're doing looks like: 
+
+```mermaid
+graph LR
+    A["Successful GitHub signin"] --> B["Get user's contributions"]
+    B-->C{"Contributor?"}
+    C-- Yes -->D["Return true"]
+    C-- No -->E["Return false"]
+```
+
+In the interest of not reinventing the wheel we opted for the [Tentacat](https://github.com/edgurgel/tentacat) library. At this point in our journey our `mix.exs` dependencies looked like this:
+
+```elixir
+  defp deps do
+    [
+      {:gettext, "~> 0.11"},
+      {:jason, "~> 1.0"},
+      {:phoenix, "~> 1.4.0"},
+      {:phoenix_html, "~> 2.11"},
+      {:plug_cowboy, "~> 2.0"},
+      {:tentacat, "~> 1.5"},
+      {:ueberauth_github, "~> 0.7.0"},
+
+      {:phoenix_live_reload, "~> 1.2", only: :dev}
+    ]
+  end
+```
+
+With our new dependency in place we can fetch (`mix deps.get`) and get on our way. Keeping our controller's simple and focused on presentation is a goal we always shoot for so we decided to implement the eligibility code in a separate module outside of the web portion of our application.
+
+We've called this new module `Registrar` in keeping with our college theme, it can be found in the `lib/admissions/registrar.ex` file.
+
+>  ### reg·is·trar
+>
+> 1. An official in a college or university who is responsible for keeping student records.
+
+Given the flow above we deteremined the best way to achieve this would be to check a list of repositories in our organization (with support for multiple organizations) for contributors who matched our user's GitHub nickname. To this end we knew we'd need to store the organization's name and its repositories. For this we opted to use a map where an organization name's is the key and the value is a list of repositories. To avoid any type casting we elected to store everything as strings, the end result of which was added to our `config.exs`:
```

Comment:
```suggestion
Given the flow above we determined the best way to achieve this would be to check a list of repositories in our organization (with support for multiple organizations) for contributors who matched our user's GitHub nickname. To this end we knew we'd need to store the organization's name and its repositories. For this we opted to use a map where an organization name's is the key and the value is a list of repositories. To avoid any type casting we elected to store everything as strings, the end result of which was added to our `config.exs`:
```